### PR TITLE
Iipb/mdbx update

### LINKED ISSUE #1014: Assertion in MDBX 0.1.8
Updating `mdbx` to 0.1.8 as #1005 did, the code randomly an assertion:
```
nimiq-client: mdbx:9268: map_resize: Assertion `size_bytes == env->me_dxb_mmap.current' failed.
```
The issue looks much more reproducible when triggering the _Devnet Release Scenarios_.
As a temporal fix 7542732 reverts it back to 0.1.7.

### PR DESCRIPTION
**Pull request by ii-cruz**
## What's in this pull request?
Updating libmdbx version. Replaced the use of cursors to the iterators now supported by mdbx. 

## Pull request checklist

- [x] All tests pass. The project builds and runs.
- [x] I have resolved any merge conflicts.
- [x] I have resolved all `clippy` and `rustfmt` warnings.

### CODE REVIEW

File: block-production/tests/mod.rs [hunk 1]
```diff
@@ -1,6 +1,7 @@
 use std::convert::TryInto;
 use std::sync::Arc;
 
+use nimiq_database::traits::WriteTransaction;
```
Comment (jsdanielh) [inline block-production/tests/mod.rs]: I'd move it below with all `nimiq_database` entries

File: blockchain/src/blockchain/blockchain.rs [hunk 2]
```diff
@@ -1,10 +1,11 @@
+use nimiq_database::traits::{Database, WriteTransaction};
```
Comment (jsdanielh) [inline blockchain/src/blockchain/blockchain.rs]: I'd move this to the `nimiq_database` entries below

File: blockchain/src/blockchain/zkp_sync.rs [hunk 3]
```diff
@@ -1,5 +1,6 @@
 use std::{cmp, mem};
 
+use nimiq_database::traits::{ReadTransaction, WriteTransaction};
```
Comment (jsdanielh) [inline blockchain/src/blockchain/zkp_sync.rs]: I'd move this include where all `nimiq` includes are

**Comment by sisou** [comment]
Also closes #1591.

**Comment by sisou** [comment]
Why the change to iterators? Are they more performant? Do they have other properties that we like?

File: blockchain/src/history/mmr_store.rs [hunk 4]
```diff
@@ -1,15 +1,15 @@
 use std::cmp;
 use std::convert::TryInto;
 
-use nimiq_database::cursor::ReadCursor;
-use nimiq_database::{Database, Transaction, WriteTransaction};
+use nimiq_database::traits::{ReadCursor, ReadTransaction, WriteTransaction};
+use nimiq_database::{TableProxy, TransactionProxy, WriteTransactionProxy};
```
Comment (jsdanielh) [inline blockchain/src/history/mmr_store.rs]:
```suggestion
use nimiq_database::{traits::{ReadCursor, ReadTransaction, WriteTransaction}, TableProxy, TransactionProxy, WriteTransactionProxy};
```

File: blockchain/tests/inherents.rs [hunk 5]
```diff
@@ -1,7 +1,8 @@
 use nimiq_account::{BlockLogger, BlockState};
 use nimiq_block::{MacroBlock, MacroBody, MacroHeader};
 use nimiq_blockchain::{Blockchain, BlockchainConfig};
-use nimiq_database::volatile::VolatileEnvironment;
+use nimiq_database::traits::WriteTransaction;
+use nimiq_database::volatile::VolatileDatabase;
```
Comment (jsdanielh) [inline blockchain/tests/inherents.rs]:
```suggestion
use nimiq_database::{traits::WriteTransaction, volatile::VolatileDatabase};
```

File: consensus/src/sync/light/sync_stream.rs [hunk 6]
```diff
@@ -424,13 +424,14 @@ mod tests {
     use std::sync::Arc;
 
     use futures::StreamExt;
+    use nimiq_database::traits::WriteTransaction;
```
Comment (jsdanielh) [inline consensus/src/sync/light/sync_stream.rs]: I'd move the include in the `nimiq-database` listing below.

File: database/src/mdbx/cursor.rs [hunk 7]
```diff
@@ -0,0 +1,256 @@
+use std::{borrow::Cow, marker::PhantomData};
+
+use libmdbx::{TransactionKind, WriteFlags, RO, RW};
+use nimiq_database_value::{AsDatabaseBytes, FromDatabaseValue};
```
Comment (jsdanielh) [inline database/src/mdbx/cursor.rs]:
```suggestion
use libmdbx::{TransactionKind, WriteFlags, RO, RW};

use nimiq_database_value::{AsDatabaseBytes, FromDatabaseValue};
```

File: database/src/mdbx/cursor.rs [hunk 8]
```diff
@@ -0,0 +1,256 @@
...
+
+use super::IntoIter;
+use crate::traits::{ReadCursor, WriteCursor};
+
+use super::DbKvPair;
```
Comment (jsdanielh) [inline database/src/mdbx/cursor.rs]: I'd merge this include into the `super` listing above

File: database/src/mdbx/cursor.rs [hunk 9]
```diff
@@ -0,0 +1,256 @@
...
+
+pub struct MdbxCursor<'txn, K: TransactionKind> {
```
Comment (jsdanielh) [inline database/src/mdbx/cursor.rs:12]: Could you add some `rustdoc`?

File: database/src/mdbx/cursor.rs [hunk 10]
```diff
@@ -0,0 +1,256 @@
...
+    cursor: libmdbx::Cursor<'txn, K>,
+}
+pub type MdbxReadCursor<'txn> = MdbxCursor<'txn, RO>;
```
Comment (jsdanielh) [inline database/src/mdbx/cursor.rs:16]: Could you add some `rustdoc`?

File: database/src/mdbx/cursor.rs [hunk 11]
```diff
@@ -0,0 +1,256 @@
...
+pub type MdbxWriteCursor<'txn> = MdbxCursor<'txn, RW>;
```
Comment (jsdanielh) [inline database/src/mdbx/cursor.rs:18]: Could you add some `rustdoc`?

**Comment by ii-cruz** [comment]
Iterators are easier to read and understand. The code also looks simpler. In terms of efficiency, we believe it to be rather similar.

File: validator/src/validator.rs [hunk 12]
```diff
@@ -10,6 +10,7 @@ use std::time::Duration;
 
 use futures::stream::{BoxStream, Stream, StreamExt};
 use linked_hash_map::LinkedHashMap;
+use nimiq_database::traits::Database;
```
Comment (jsdanielh) [inline validator/src/validator.rs]: I'd move the include to the `nimiq-database` listing below

File: rpc-server/src/dispatchers/wallet.rs [hunk 13]
```diff
@@ -1,6 +1,7 @@
 use std::sync::Arc;
 
 use async_trait::async_trait;
+use nimiq_database::traits::WriteTransaction;
```
Comment (jsdanielh) [inline rpc-server/src/dispatchers/wallet.rs]: I'd move the include to the `nimiq` includes below

File: primitives/account/tests/staking_contract.rs [hunk 14]
```diff
@@ -7,7 +7,8 @@ use nimiq_bls::{
     CompressedPublicKey as BlsPublicKey, KeyPair as BlsKeyPair, SecretKey as BlsSecretKey,
 };
 use nimiq_collections::BitSet;
-use nimiq_database::{volatile::VolatileEnvironment, WriteTransaction};
+use nimiq_database::traits::Database;
+use nimiq_database::volatile::VolatileDatabase;
```
Comment (jsdanielh) [inline primitives/account/tests/staking_contract.rs]:
```suggestion
use nimiq_database::{traits::Database, volatile::VolatileDatabase};
```

File: primitives/account/src/data_store.rs [hunk 15]
```diff
@@ -100,17 +100,17 @@ mod tests {
     use crate::data_store::DataStore;
     use crate::data_store_ops::DataStoreReadOps;
     use crate::AccountsTrie;
-    use nimiq_database::volatile::VolatileEnvironment;
-    use nimiq_database::{ReadTransaction, WriteTransaction};
+    use nimiq_database::traits::{Database, WriteTransaction};
+    use nimiq_database::volatile::VolatileDatabase;
```
Comment (jsdanielh) [inline primitives/account/src/data_store.rs]:
```suggestion
    use nimiq_database::{traits::{Database, WriteTransaction}, volatile::VolatileDatabase};
```

File: genesis-builder/src/lib.rs [hunk 16]
```diff
@@ -6,6 +6,7 @@ use std::fs::{read_to_string, OpenOptions};
 use std::io::Error as IoError;
 use std::path::Path;
 
+use nimiq_database::traits::{Database, WriteTransaction};
```
Comment (jsdanielh) [inline genesis-builder/src/lib.rs]: I'd move this include to the `nimiq-database` listing below

File: blockchain/src/chain_store.rs [hunk 17]
```diff
@@ -280,65 +286,51 @@ impl ChainStore {
...
-                Some((_, hash)) => hash,
-                None => break,
-            };
+        let cursor = txn.cursor(&self.height_idx);
+        let iter = cursor.into_iter_dup_of(&block_height).map(|(_, hash)| hash);
+
+        for block_hash in iter {
+            blocks.push(
+                self.get_block(&block_hash, include_body, Some(txn))
+                    .expect("Corrupted store: Block referenced from index not found"),
```
Comment (jsdanielh) [inline blockchain/src/chain_store.rs]: Maybe include the block hash as well?

File: blockchain/src/history/history_store.rs [hunk 18]
```diff
@@ -2,19 +2,24 @@ use std::cmp;
 use std::collections::{HashSet, VecDeque};
 
 use beserial::Serialize;
-use nimiq_database::cursor::{ReadCursor, WriteCursor};
+use nimiq_database::traits::{Database, ReadTransaction, WriteTransaction};
```
Comment (jsdanielh) [inline blockchain/src/history/history_store.rs]: Maybe collapse both `nimiq-database` includes?

File: blockchain/src/history/history_store.rs [hunk 19]
```diff
@@ -31,20 +36,20 @@ use crate::history::{mmr_store::MMRStore, ordered_hash::OrderedHash, HistoryTree
 /// only has macro block headers) that that given transaction happened.
 #[derive(Debug)]
 pub struct HistoryStore {
-    env: Environment,
+    env: DatabaseProxy,
     // A database of all history trees indexed by their epoch number.
```
Comment (jsdanielh) [inline blockchain/src/history/history_store.rs]: A table maybe? And also, could you convert it to `rustdoc`?

File: blockchain/src/history/history_store.rs [hunk 20]
```diff
@@ -31,20 +36,20 @@ use crate::history::{mmr_store::MMRStore, ordered_hash::OrderedHash, HistoryTree
...
-    hist_tree_db: Database,
+    hist_tree_db: TableProxy,
     // A database of all extended transactions indexed by their hash (= leaf hash in the history
```
Comment (jsdanielh) [inline blockchain/src/history/history_store.rs]: A table maybe? And also, could you convert it to `rustdoc`?

File: blockchain/src/history/history_store.rs [hunk 21]
```diff
@@ -31,20 +36,20 @@ use crate::history::{mmr_store::MMRStore, ordered_hash::OrderedHash, HistoryTree
...
     // tree).
-    ext_tx_db: Database,
+    ext_tx_db: TableProxy,
     // A database of all leaf hashes and indexes indexed by the hash of the transaction. This way we
```
Comment (jsdanielh) [inline blockchain/src/history/history_store.rs]: A table maybe? And also, could you convert it to `rustdoc`?

File: blockchain/src/history/history_store.rs [hunk 22]
```diff
@@ -31,20 +36,20 @@ use crate::history::{mmr_store::MMRStore, ordered_hash::OrderedHash, HistoryTree
...
     // can start with a transaction hash and find it in the MMR.
-    tx_hash_db: Database,
+    tx_hash_db: TableProxy,
     // A database of the last leaf index for each block number.
```
Comment (jsdanielh) [inline blockchain/src/history/history_store.rs]: A table maybe? And also, could you convert it to `rustdoc`?

File: blockchain/src/history/history_store.rs [hunk 23]
```diff
@@ -31,20 +36,20 @@ use crate::history::{mmr_store::MMRStore, ordered_hash::OrderedHash, HistoryTree
...
-    last_leaf_db: Database,
+    last_leaf_db: TableProxy,
     // A database of all transaction (and reward inherent) hashes indexed by their sender and
```
Comment (jsdanielh) [inline blockchain/src/history/history_store.rs]: A table maybe? And also, could you convert it to `rustdoc`?

File: database/src/lib.rs [hunk 24]
```diff
@@ -1,58 +1,17 @@
...
 
-use crate::cursor::{ReadCursor, WriteCursor as WriteCursorTrait};
-
-#[macro_use]
-pub mod cursor;
+mod error;
 pub mod mdbx;
+pub mod proxy;
+pub mod traits;
 pub mod volatile;
```
Comment (jsdanielh) [inline database/src/lib.rs:9]: Could you add some `rustdoc` for the public modules?

File: database/src/mdbx/database.rs [hunk 25]
```diff
@@ -0,0 +1,179 @@
...
+    type WriteTransaction<'db> = MdbxWriteTransaction<'db>
+    where
+        Self: 'db;
+
+    fn open_table(&self, name: String) -> Self::Table {
+        self.open_table_with_flags(name, TableFlags::empty())
+    }
+
+    fn open_table_with_flags(&self, name: String, flags: TableFlags) -> Self::Table {
+        // This is an implicit transaction, so take the lock first.
```
Comment (jsdanielh) [inline database/src/mdbx/database.rs:33]:
```suggestion
        // This is an implicit transaction, so take the lock first
```

File: database/src/mdbx/database.rs [hunk 26]
```diff
@@ -0,0 +1,179 @@
...
+        let mut table_flags = libmdbx::TableFlags::CREATE;
+
+        // Translate flags.
```
Comment (jsdanielh) [inline database/src/mdbx/database.rs:36]:
```suggestion
        // Translate flags
```

File: database/src/mdbx/database.rs [hunk 27]
```diff
@@ -0,0 +1,179 @@
...
+use log::info;
+
+use crate::{traits::Database, DatabaseProxy, Error, TableFlags};
+
+use super::{MdbxReadTransaction, MdbxWriteTransaction};
+
+pub(super) type DbKvPair<'a> = (Cow<'a, [u8]>, Cow<'a, [u8]>);
+
+#[derive(Clone, Debug)]
+pub struct MdbxDatabase {
```
Comment (jsdanielh) [inline database/src/mdbx/database.rs:15]: Could you add some `rustdoc`?

File: database/src/mdbx/database.rs [hunk 28]
```diff
@@ -0,0 +1,179 @@
...
+            no_rdahead: true,
+            mode: libmdbx::Mode::ReadWrite {
+                sync_mode: libmdbx::SyncMode::Durable,
+            },
+            ..Default::default()
+        };
+
+        db.set_flags(db_flags);
+
+        // This is only required if multiple databases will be used in the environment.
```
Comment (jsdanielh) [inline database/src/mdbx/database.rs]:
```suggestion
        // This is only required if multiple databases will be used in the environment
```

Comment (jsdanielh) [reply database/src/mdbx/database.rs]: And we no longer have the concept of environment/databases, do we? Maybe tables in databases now?

File: database/src/mdbx/database.rs [hunk 29]
```diff
@@ -0,0 +1,179 @@
...
+        db.set_max_tables(max_tables as usize);
+        if let Some(max_readers) = max_readers {
+            db.set_max_readers(max_readers);
+        }
+
+        let db = db.open(path)?;
+
+        let info = db.info()?;
+        let cur_mapsize = info.map_size();
+        info!("MDBX memory map size: {}", cur_mapsize);
```
Comment (jsdanielh) [inline database/src/mdbx/database.rs]: Use tracing style:
```suggestion
        info!(cur_mapsize, "MDBX memory map size");
```

File: database/src/mdbx/database.rs [hunk 30]
```diff
@@ -0,0 +1,179 @@
...
+
+        let mdbx = MdbxDatabase { db: Arc::new(db) };
+        if mdbx.need_resize(0) {
+            info!("MDBX memory needs to be resized.");
```
Comment (jsdanielh) [inline database/src/mdbx/database.rs:137]:
```suggestion
            info!("MDBX memory needs to be resized");
```

File: database/src/mdbx/database.rs [hunk 31]
```diff
@@ -0,0 +1,179 @@
...
+        let stat = self.db.stat().unwrap();
+
+        let size_used = (stat.page_size() as usize) * (info.last_pgno() + 1);
+
+        if threshold_size > 0 && info.map_size() - size_used < threshold_size {
+            info!("DB resize (threshold-based)");
+            info!("DB map size: {}", info.map_size());
+            info!("Space used: {}", size_used);
+            info!("Space remaining: {}", info.map_size() - size_used);
+            info!("Size threshold: {}", threshold_size);
```
Comment (jsdanielh) [inline database/src/mdbx/database.rs]:
```suggestion
            info!( threshold-based, map_size=info.map_size(), size_used, space_remaining=info.map_size() - size_used, threshold_size, "DB settings:");
```

File: database/src/mdbx/database.rs [hunk 32]
```diff
@@ -0,0 +1,179 @@
...
+
+        if (size_used as f64) / (info.map_size() as f64) > resize_percent {
+            info!("DB resize (percent-based)");
+            info!("DB map size: {}", info.map_size());
+            info!("Space used: {}", size_used);
+            info!("Space remaining: {}", info.map_size() - size_used);
+            info!(
+                "Percent used: {:.2}",
+                (size_used as f64) / (info.map_size() as f64)
+            );
```
Comment (jsdanielh) [inline database/src/mdbx/database.rs]:
```suggestion
            info!(map_size=info.map_size(), size_used, space_remaining = info.map_size() - size_used, percent_used=                (size_used as f64) / (info.map_size() as f64), "DB resize (percent-based)");
```

File: database/src/mdbx/database.rs [hunk 33]
```diff
@@ -0,0 +1,179 @@
...
+            return true;
+        }
+
+        false
+    }
+}
+
+#[derive(Debug)]
+pub struct MdbxTable {
```
Comment (jsdanielh) [inline database/src/mdbx/database.rs:182]: Could you add `rustdoc` to it?

File: database/src/mdbx/iterators.rs [hunk 34]
```diff
@@ -0,0 +1,28 @@
+use std::{borrow::Cow, marker::PhantomData};
+
+use libmdbx::TransactionKind;
+
+use nimiq_database_value::FromDatabaseValue;
+
+pub struct IntoIter<'txn, Kind: TransactionKind, K: FromDatabaseValue, V: FromDatabaseValue> {
```
Comment (jsdanielh) [inline database/src/mdbx/iterators.rs:9]: Could you add `rustdoc` to it?

File: database/src/mdbx/mod.rs [hunk 35]
```diff
@@ -0,0 +1,238 @@
...
+
+            // Remove smaller value and write larger value.
+            let mut txw = db.write_transaction();
+            assert_eq!(txw.get::<str, u32>(&table, "test"), Some(12));
+            txw.remove_item::<str, u32>(&table, "test", &12);
+            txw.put::<str, u32>(&table, "test", &5783);
+            assert_eq!(txw.get::<str, u32>(&table, "test"), Some(125));
+            txw.commit();
+
+            // Have a new ReadTransaction read the smallest value.
```
Comment (jsdanielh) [inline database/src/mdbx/mod.rs]:
```suggestion
            // Have a new ReadTransaction read the smaller value.
```

File: database/src/mdbx/mod.rs [hunk 36]
```diff
@@ -0,0 +1,238 @@
...
+
+            assert_eq!(
+                cursor.get_current::<String, u32>(),
+                Some((test1.clone(), 5783))
+            );
+
+            assert_eq!(cursor.get_current::<String, u32>(), Some((test1, 5783)));
+            assert!(cursor.prev_no_duplicate::<String, u32>().is_none());
+            assert_eq!(cursor.next::<String, u32>(), Some((test2, 5783)));
+            //            assert_eq!(cursor.seek_range_key::<String, u32>("test"), Some((test1.clone(), 12)));
```
Comment (jsdanielh) [inline database/src/mdbx/mod.rs]: Maybe remove it?

File: database/src/mdbx/transaction.rs [hunk 37]
```diff
@@ -0,0 +1,140 @@
+use std::borrow::Cow;
+
+use libmdbx::{NoWriteMap, TransactionKind, WriteFlags, RO, RW};
+use nimiq_database_value::{AsDatabaseBytes, FromDatabaseValue, IntoDatabaseValue};
```
Comment (jsdanielh) [inline database/src/mdbx/transaction.rs]:
```suggestion
use libmdbx::{NoWriteMap, TransactionKind, WriteFlags, RO, RW};

use nimiq_database_value::{AsDatabaseBytes, FromDatabaseValue, IntoDatabaseValue};
```

File: database/src/mdbx/transaction.rs [hunk 38]
```diff
@@ -0,0 +1,140 @@
...
+
+use super::{MdbxCursor, MdbxTable, MdbxWriteCursor};
+use crate::traits::{ReadTransaction, WriteTransaction};
+
+#[derive(Debug)]
+pub struct MdbxTransaction<'db, K: TransactionKind> {
```
Comment (jsdanielh) [inline database/src/mdbx/transaction.rs:12]: Maybe add some `rustdoc`?

File: database/src/mdbx/transaction.rs [hunk 39]
```diff
@@ -0,0 +1,140 @@
...
+    txn: libmdbx::Transaction<'db, K, NoWriteMap>,
+}
+pub type MdbxReadTransaction<'db> = MdbxTransaction<'db, RO>;
```
Comment (jsdanielh) [inline database/src/mdbx/transaction.rs:16]: Maybe add some `rustdoc`?

File: database/src/mdbx/transaction.rs [hunk 40]
```diff
@@ -0,0 +1,140 @@
...
+pub type MdbxWriteTransaction<'db> = MdbxTransaction<'db, RW>;
```
Comment (jsdanielh) [inline database/src/mdbx/transaction.rs:18]: Maybe add some `rustdoc`?

File: database/src/proxy/cursor.rs [hunk 41]
```diff
@@ -0,0 +1,224 @@
...
+
+use nimiq_database_value::{AsDatabaseBytes, FromDatabaseValue};
+
+use crate::{
+    mdbx::{IntoIter, MdbxReadCursor, MdbxWriteCursor},
+    traits::ReadCursor,
+};
+
+#[derive(Clone)]
+pub enum CursorProxy<'txn> {
```
Comment (jsdanielh) [inline database/src/proxy/cursor.rs:12]: Maybe add some `rustdoc`?

File: database/src/proxy/cursor.rs [hunk 42]
```diff
@@ -0,0 +1,224 @@
...
+        match self {
+            CursorProxy::ReadCursor(cursor) => IntoIterProxy::ReadIter(cursor.into_iter_from(key)),
+            CursorProxy::WriteCursor(cursor) => {
+                IntoIterProxy::WriteIter(cursor.into_iter_from(key))
+            }
+        }
+    }
+}
+
+pub enum IntoIterProxy<'txn, K: FromDatabaseValue, V: FromDatabaseValue> {
```
Comment (jsdanielh) [inline database/src/proxy/cursor.rs:212]: Maybe add some `rustdoc`?

File: database/src/proxy/database.rs [hunk 43]
```diff
@@ -0,0 +1,53 @@
+use crate::{
+    mdbx, traits::Database, volatile, TableProxy, TransactionProxy, WriteTransactionProxy,
+};
+
+#[derive(Clone, Debug)]
+pub enum DatabaseProxy {
```
Comment (jsdanielh) [inline database/src/proxy/database.rs:7]: Maybe add some `rustdoc`?

File: database/src/proxy/mod.rs [hunk 44]
```diff
@@ -0,0 +1,18 @@
...
+pub use cursor::*;
+pub use database::*;
+pub use transaction::*;
+
+use crate::{
+    mdbx::MdbxTable,
+    traits::{ReadTransaction, WriteTransaction},
+};
+
+pub type TableProxy = MdbxTable;
```
Comment (jsdanielh) [inline database/src/proxy/mod.rs:12]: Maybe add some `rustdoc`?

File: database/src/proxy/mod.rs [hunk 45]
```diff
@@ -0,0 +1,18 @@
...
+
+pub type Cursor<'txn> = <TransactionProxy<'txn> as ReadTransaction<'txn>>::Cursor<'txn>;
```
Comment (jsdanielh) [inline database/src/proxy/mod.rs]: Maybe add some `rustdoc`?

File: database/src/proxy/mod.rs [hunk 46]
```diff
@@ -0,0 +1,18 @@
...
+pub type WriteCursor<'txn> =
```
Comment (jsdanielh) [inline database/src/proxy/mod.rs]: Maybe add some `rustdoc`?

File: database/src/proxy/transaction.rs [hunk 47]
```diff
@@ -0,0 +1,152 @@
+use std::ops::Deref;
+
+use crate::{
+    mdbx::{MdbxReadTransaction, MdbxWriteCursor, MdbxWriteTransaction},
+    traits::{ReadTransaction, WriteTransaction},
+    CursorProxy, TableProxy,
+};
+
+#[derive(Debug)]
+pub enum TransactionProxy<'db> {
```
Comment (jsdanielh) [inline database/src/proxy/transaction.rs:12]: Maybe add some `rustdoc`?

File: database/src/proxy/transaction.rs [hunk 48]
```diff
@@ -0,0 +1,152 @@
...
+    ReadTransaction(MdbxReadTransaction<'db>),
+    WriteTransaction(MdbxWriteTransaction<'db>),
+}
+
+#[derive(Debug)]
+pub struct WriteTransactionProxy<'db> {
```
Comment (jsdanielh) [inline database/src/proxy/transaction.rs:19]: Maybe add some `rustdoc`?

File: database/src/traits/cursor.rs [hunk 49]
```diff
@@ -0,0 +1,92 @@
+use nimiq_database_value::{AsDatabaseBytes, FromDatabaseValue};
+
+pub trait ReadCursor<'txn>: Clone {
```
Comment (jsdanielh) [inline database/src/traits/cursor.rs:7]: Maybe add some `rustdoc`?

File: database/src/traits/cursor.rs [hunk 50]
```diff
@@ -0,0 +1,92 @@
...
+        K: AsDatabaseBytes + FromDatabaseValue,
+        V: FromDatabaseValue;
+
+    fn into_iter_from<K, V>(self, key: &K) -> Self::IntoIter<K, V>
+    where
+        K: AsDatabaseBytes + FromDatabaseValue,
+        V: FromDatabaseValue;
+}
+
+pub trait WriteCursor<'txn>: ReadCursor<'txn> {
```
Comment (jsdanielh) [inline database/src/traits/cursor.rs:98]: Maybe add some `rustdoc`?

File: database/src/traits/database.rs [hunk 51]
```diff
@@ -0,0 +1,18 @@
+use super::{ReadTransaction, WriteTransaction};
+use crate::TableFlags;
+
+pub trait Database: Sized {
```
Comment (jsdanielh) [inline database/src/traits/database.rs:5]: Maybe add some `rustdoc`?

File: database/src/traits/transaction.rs [hunk 52]
```diff
@@ -0,0 +1,59 @@
+use nimiq_database_value::{AsDatabaseBytes, FromDatabaseValue, IntoDatabaseValue};
+
+use super::{ReadCursor, WriteCursor};
+
+pub trait ReadTransaction<'db>: Sized {
```
Comment (jsdanielh) [inline database/src/traits/transaction.rs:6]: Maybe add some `rustdoc`?

File: database/src/traits/transaction.rs [hunk 53]
```diff
@@ -0,0 +1,59 @@
...
+
+    fn get<K, V>(&self, table: &Self::Table, key: &K) -> Option<V>
+    where
+        K: AsDatabaseBytes + ?Sized,
+        V: FromDatabaseValue;
+
+    fn cursor<'txn>(&'txn self, table: &Self::Table) -> Self::Cursor<'txn>;
+}
+
+pub trait WriteTransaction<'db>: ReadTransaction<'db> + Sized {
```
Comment (jsdanielh) [inline database/src/traits/transaction.rs:23]: Maybe add some `rustdoc`?

File: database/src/volatile.rs [hunk 54]
```diff
@@ -4,567 +4,275 @@ use tempfile::TempDir;
 
 use super::mdbx::*;
 use super::*;
-use crate::cursor::{ReadCursor, WriteCursor as WriteCursorTrait};
+use crate::traits::Database;
 
 #[derive(Debug)]
-pub struct VolatileEnvironment {
+pub struct VolatileDatabase {
```
Comment (jsdanielh) [inline database/src/volatile.rs:11]: Maybe add some `rustdoc`?

File: blockchain/src/history/mmr_store.rs [hunk 55]
```diff
@@ -51,27 +53,27 @@ impl<'a, 'env> MMRStore<'a, 'env> {
...
-        hist_tree_db: &'a TableProxy,
+        hist_tree_table: &'a TableProxy,
         tx: &'a mut WriteTransactionProxy<'env>,
         epoch_number: u32,
     ) -> Self {
-        let size = Self::get_size(hist_tree_db, tx, epoch_number);
+        let size = Self::get_size(hist_tree_table, tx, epoch_number);
         MMRStore {
-            hist_tree_db,
+            hist_tree_table: hist_tree_table,
```
Comment (jsdanielh) [inline blockchain/src/history/mmr_store.rs]: Could you fix this clippy error?

File: blockchain/src/chain_store.rs [hunk 56]
```diff
@@ -329,7 +330,10 @@ impl ChainStore {
         for block_hash in iter {
             blocks.push(
                 self.get_block(&block_hash, include_body, Some(txn))
-                    .expect("Corrupted store: Block referenced from index not found"),
+                    .expect(&format!(
+                        "Corrupted store: Block {} referenced from index not found",
+                        block_hash,
+                    )),
```
Comment (jsdanielh) [inline blockchain/src/chain_store.rs]: Could you fix this clippy warning?

**Review by jsdanielh — CHANGES_REQUESTED** [review]

**Review by jsdanielh — APPROVED** [review]

**Pull request merged**